File: Client/Linux/lib/Management/CompanyManagementClient.cpp
#include "ManagementClient.h"
#include "../RestConnection.h"
#include "common.h"
#include "ArduinoJson.h"

namespace Thriot { namespace Management {

/**
Create a new instance of the company management client

@param restConnection Class providing REST function calls

*/
CompanyManagementClient::CompanyManagementClient(RestConnection* restConnection)
{
	_restConnection = restConnection;
}

/**
List all companies that the currently logged in user has permission to.
The function returns an empty list on error.

@return List of companies*/
vector<Small> CompanyManagementClient::List()
{
	vector<Small> companies;
	Response httpResponse = _restConnection->Get("companies");

	if(httpResponse.Code!= 200)
		return companies;

	DynamicJsonBuffer jsonBufferResponse;
	JsonArray& respJson = jsonBufferResponse.parseArray((char *)httpResponse.Body.c_str());

	for(JsonArray::iterator element = respJson.begin(); element!= respJson.end(); ++element)
	{
		Small small;
		small.Id = (*element)["Id"].asString();
		small.Name = (*element)["Name"].asString();
		companies.push_back(small);
	}

	return companies;
}

/**
Get a single company by it's unique identifier.
The function returns an uninitialized entity on error.

@param id Company unique identifier

@return A company entity */
Company CompanyManagementClient::Get(const string& id)
{
	Company company;

	Response httpResponse = _restConnection->Get(string("companies/") + id);

	if(httpResponse.Code!= 200)
		return company;

	DynamicJsonBuffer jsonBufferResponse;
	JsonObject& respJson = jsonBufferResponse.parseObject((char *)httpResponse.Body.c_str());
	
	company.Id = string(respJson["Id"].asString());
	company.Name = string(respJson["Name"].asString());

	ParseTelemetryDataSinkSettings(respJson, company.TelemetryDataSinkSettings);

	return company;
}

/**
Create a new company. The function returns an empty string on error.

@param company Company parameters

@return 32-characters long unique identifier of the company */
string CompanyManagementClient::Create(const Company& company)
{
	DynamicJsonBuffer jsonBufferRequest;
	JsonObject& createJson = jsonBufferRequest.createObject();

	createJson["Name"] = company.Name.c_str();

	char jsonBuffer[256];
	createJson.printTo(jsonBuffer, sizeof(jsonBuffer));

	Response httpResponse =_restConnection->Post("companies", "application/json",jsonBuffer);

	if(httpResponse.Code != 200)
		return "";

	string id = httpResponse.Body;
	StripQuotes(id);

	return id;
}

/**
Update a company. On error return the http status code otherwise 0.

@param company Company to update

@return status code */
int CompanyManagementClient::Update(const Company& company)
{
	DynamicJsonBuffer jsonBufferRequest;
	JsonObject& updateJson = jsonBufferRequest.createObject();

	updateJson["Id"] = company.Id.c_str();
	updateJson["Name"] = company.Name.c_str();

	char jsonBuffer[256];
	updateJson.printTo(jsonBuffer, sizeof(jsonBuffer));

	Response httpResponse =_restConnection->Put("companies", "application/json", jsonBuffer);

	if(httpResponse.Code != 204)
		return httpResponse.Code;

	return 0;
}

/**
Delete a company specified by id. On error return the http status code otherwise 0.

@param id Company specified by id to delete

@return status code*/
int CompanyManagementClient::Delete(const string& id)
{
	Response httpResponse =_restConnection->Delete("companies/" + id);

	if(httpResponse.Code != 204)
		return httpResponse.Code;

	return 0;
}

/**
List services that are set up under the company. On error the function returns an empty list.

@param id Company id

@return List of services*/
vector<Small> CompanyManagementClient::ListServices(const string& id)
{
	vector<Small> services;
	Response httpResponse = _restConnection->Get("companies/" + id + "/services");

	if(httpResponse.Code!= 200)
		return services;

	DynamicJsonBuffer jsonBufferResponse;
	JsonArray& respJson = jsonBufferResponse.parseArray((char *)httpResponse.Body.c_str());

	for(JsonArray::iterator element = respJson.begin(); element!= respJson.end(); ++element)
	{
		Small small;
		small.Id = (*element)["Id"].asString();
		small.Name = (*element)["Name"].asString();
		services.push_back(small);
	}

	return services;
}

/**
Updates the telemetry data sinks specified for a company.
On error the functions returns the http status code otherwise 0.

@param id Company identifier
@param telemetryDataSinkParameters Collection of parameterized telemetry data sinks

@return status code */
int CompanyManagementClient::UpdateIncomingTelemetryDataSinks(const string& id, const vector<TelemetryDataSinkParameters>& telemetryDataSinkParameters)
{
	string tdspJsonString = BuildTelmetryDataSinkParametersListJson(telemetryDataSinkParameters);

	Response httpResponse = _restConnection->Post("companies/" + id + "/incomingTelemetryDataSinks", "application/json", tdspJsonString);

	if(httpResponse.Code!= 204)
		return httpResponse.Code;

	return 0;
}

/**
List users who have access to the company.
On error an empty list will be returned.

@param id Id of the company

@return List of users */
vector<SmallUser> CompanyManagementClient::ListUsers(const string& id)
{
	vector<SmallUser> services;
	Response httpResponse = _restConnection->Get("companies/" + id + "/users");

	if(httpResponse.Code!= 200)
		return services;

	DynamicJsonBuffer jsonBufferResponse;
	JsonArray& respJson = jsonBufferResponse.parseArray((char *)httpResponse.Body.c_str());

	for(JsonArray::iterator element = respJson.begin(); element!= respJson.end(); ++element)
	{
		SmallUser small;
		small.Id = (*element)["Id"].asString();
		small.Name = (*element)["Name"].asString();
		services.push_back(small);
	}

	return services;
}

/**
Add users to the company.
Http status code on error, otherwise 0.

@param companyId Company identifier
@param userId User identifier

@return status code*/
int CompanyManagementClient::AddUser(const string& companyId, const string& userId)
{
	DynamicJsonBuffer jsonBufferRequest;
	JsonObject& updateJson = jsonBufferRequest.createObject();

	updateJson["CompanyId"] = companyId.c_str();
	updateJson["UserId"] = userId.c_str();

	char jsonBuffer[256];
	updateJson.printTo(jsonBuffer, sizeof(jsonBuffer));

	Response httpResponse =_restConnection->Post("companies/addUser", "application/json", jsonBuffer);

	if(httpResponse.Code != 204)
		return httpResponse.Code;

	return 0;
}
}}

File: Client/DotNet/IoT.Client.DotNet/Management/CompanyManagementClient.cs
using System.Collections.Generic;
using System.Net;

namespace Thriot.Client.DotNet.Management
{
    /// <summary>
    /// This class is responsible for managing companies and strictly company-related artifacts
    /// </summary>
    public class CompanyManagementClient : SpecificManagementClient
    {
        internal CompanyManagementClient(IRestConnection restConnection) : base(restConnection)
        {
        }

        /// <summary>
        /// List all companies that the currently logged in user can see and access
        /// 
        /// Send GET request to the APIROOT/companies Url
        /// </summary>
        /// <returns>List of the companies with the most important properties (id, name)</returns>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public IEnumerable<Small> List()
        {
            var items = RestConnection.Get("companies");

            return JsonSerializer.Deserialize<List<Small>>(items);
        }

        /// <summary>
        /// Get the full detail of a company
        /// 
        /// Send GET request to the APIROOT/companies/id Url
        /// </summary>
        /// <param name="id">Unique identifier of a company</param>
        /// <returns>Company details</returns>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public Company Get(string id)
        {
            var response = RestConnection.Get("companies/" + id);

            return JsonSerializer.Deserialize<Company>(response);
        }

        /// <summary>
        /// Create a new company
        /// 
        /// Send POST request to the APIROOT/companies Url
        /// </summary>
        /// <param name="company">New company instance</param>
        /// <returns>Unique id of the company</returns>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public string Create(Company company)
        {
            var response = RestConnection.Post("companies", JsonSerializer.Serialize(company));

            return JsonSerializer.Deserialize<string>(response);
        }

        /// <summary>
        /// Update company. This method is used to update basic properties of the company (like name). 
        /// To update the telemetry data sink please use the <seealso cref="UpdateIncomingTelemetryDataSinks"/> method.
        /// 
        /// Send PUT request to the APIROOT/companies Url
        /// </summary>
        /// <param name="company">Company instance</param>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public void Update(Company company)
        {
            RestConnection.Put("companies", JsonSerializer.Serialize(company));
        }

        /// <summary>
        /// Delete the company with the given id. The company must be empty.
        /// 
        /// Send DELETE request to the APIROOT/companies/id Url
        /// </summary>
        /// <param name="id">Unique identifier of the company</param>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public void Delete(string id)
        {
            RestConnection.Delete("companies/" + id);
        }

        /// <summary>
        /// List services that are run by the company
        /// 
        /// Send GET request to the APIROOT/companies/id/services Url
        /// </summary>
        /// <param name="id">Unique identifier of the company</param>
        /// <returns>List of services</returns>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public IEnumerable<Small> ListServices(string id)
        {
            var items = RestConnection.Get("companies/" + id + "/services");

            return JsonSerializer.Deserialize<List<Small>>(items);
        }

        /// <summary>
        /// List of users who can see and have access to the company
        /// 
        /// Send GET request to the APIROOT/companies/id/users Url
        /// </summary>
        /// <param name="id">Unique identifier of the company</param>
        /// <returns>List of users</returns>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public IEnumerable<SmallUser> ListUsers(string id)
        {
            var items = RestConnection.Get("companies/" + id + "/users");

            return JsonSerializer.Deserialize<List<SmallUser>>(items);
        }

        /// <summary>
        /// Add user to the company to provide access to the company for a given user.
        /// Use the <see cref="UserManagementClient.FindUser"/> method to find the user by email address.
        /// 
        /// Send POST request to the APIROOT/companies/adduser Url
        /// </summary>
        /// <param name="companyUser">Company-user pair</param>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public void AddUser(CompanyUser companyUser)
        {
            RestConnection.Post("companies/adduser", JsonSerializer.Serialize(companyUser));
        }

        /// <summary>
        /// Update the telemetry data sinks for the company. These telemetry data sinks can be overriden at service level by the <see cref="ServiceManagementClient.UpdateIncomingTelemetryDataSinks"/> method and
        /// at network level by the <see cref="NetworkManagementClient.UpdateIncomingTelemetryDataSinks"/> method. If there is no override then all the devices under the company will use the telemetry data
        /// sinks configured by this method.
        /// 
        /// Send POST request to the APIROOT/companies/id/incomingTelmetryDataSinks Url
        /// </summary>
        /// <param name="id">Unique identifier of the company</param>
        /// <param name="telemetryDataSinkParameters">List of telemetry data sinks</param>
        /// <exception cref="WebException">This exception indicates some service level error. Please refer to the HTTP error code for more information</exception>
        public void UpdateIncomingTelemetryDataSinks(string id, IEnumerable<TelemetryDataSinkParameters> telemetryDataSinkParameters)
        {
            RestConnection.Post("companies/" + id + "/incomingTelemetryDataSinks", JsonSerializer.Serialize(telemetryDataSinkParameters, true));
        }
    }
}
